mcs/class/System/Mono.AppleTls/AppleTlsContext.cs

```diff
@@ -0,0 +1,931 @@
+#else
+using Mono.Security.Interface;
+#endif
+
+using Mono.Net;
+using Mono.Net.Security;
+using Mono.Util;
+
+using ObjCRuntime;
+
+using nint = System.IntPtr;
+
+namespace Mono.AppleTls
+{
+	class AppleTlsContext : MobileTlsContext
+	{
+		GCHandle handle;
+		IntPtr context;
+		IntPtr connectionId;
+		SslReadFunc readFunc;
+		SslWriteFunc writeFunc;
+
+		SecIdentity serverIdentity;
+		SecIdentity clientIdentity;
+
+		X509Certificate remoteCertificate;
+		X509Certificate localClientCertificate;
+		MonoTlsConnectionInfo connectionInfo;
+		bool havePeerTrust;
+		bool isAuthenticated;
+		bool handshakeFinished;
+		int handshakeStarted;
+
+		bool closed;
+		bool disposed;
+		bool closedGraceful;
+		int pendingIO;
+
+		Exception lastException;
+
+		public AppleTlsContext (
+			MobileAuthenticatedStream parent, bool serverMode, string targetHost,
+			SSA.SslProtocols enabledProtocols, X509Certificate serverCertificate,
+			X509CertificateCollection clientCertificates, bool askForClientCert)
+			: base (parent, serverMode, targetHost, enabledProtocols,
+				serverCertificate, clientCertificates, askForClientCert)
+		{
+			handle = GCHandle.Alloc (this);
+			connectionId = GCHandle.ToIntPtr (handle);
+			readFunc = NativeReadCallback;
+			writeFunc = NativeWriteCallback;
+
+			if (IsServer) {
+				if (serverCertificate == null)
+					throw new ArgumentNullException ("serverCertificate");
+			}
+		}
+
+		public IntPtr Handle {
+			get {
+				if (!HasContext)
+					throw new ObjectDisposedException ("AppleTlsContext");
+				return context;
+			}
+		}
+
+		public override bool HasContext {
+			get { return !disposed && context != IntPtr.Zero; }
+		}
+
+		[SD.Conditional ("APPLE_TLS_DEBUG")]
+		protected void Debug (string message, params object[] args)
+		{
+			Console.Error.WriteLine ("MobileTlsStream({0}): {1}", Parent.ID, string.Format (message, args));
+		}
+
+		void CheckStatusAndThrow (SslStatus status, params SslStatus[] acceptable)
+		{
+			var last = Interlocked.Exchange (ref lastException, null);
+			if (last != null)
+				throw last;
+
+			if (status == SslStatus.Success || Array.IndexOf (acceptable, status) > -1)
+				return;
+
+			switch (status) {
+			case SslStatus.ClosedAbort:
+				throw new IOException ("Connection closed.");
+
+			case SslStatus.BadCert:
+				throw new TlsException (AlertDescription.BadCertificate);
+
+			case SslStatus.UnknownRootCert:
+			case SslStatus.NoRootCert:
+			case SslStatus.XCertChainInvalid:
+				throw new TlsException (AlertDescription.CertificateUnknown, status.ToString ());
+
+			case SslStatus.CertExpired:
+			case SslStatus.CertNotYetValid:
+				throw new TlsException (AlertDescription.CertificateExpired);
+
+			case SslStatus.Protocol:
+				throw new TlsException (AlertDescription.ProtocolVersion);
+
+			default:
+				throw new TlsException (AlertDescription.InternalError, "Unknown Secure Transport error `{0}'.", status);
+			}
+		}
+
+		void SetException (string message, Exception exception)
```

Comment:
It should actually be used everywhere as it should provide better exceptions - not sure why we're not doing that.